File: app/src/main/java/com/wxm/camerajob/base/utility/FileLogger.java
package com.wxm.camerajob.base.utility;

import android.os.Environment;

import java.io.File;
import java.io.IOException;
import java.util.logging.FileHandler;
import java.util.logging.Level;
import java.util.logging.LogRecord;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

/**
 * 写在文件中的日志
 * Created by 123 on 2016/6/18.
 */
public class FileLogger {
    public final static String LOG_NAME = "camerajob_run_%g.log";

    private String mLogTag;
    public Logger mLoger;
    private FileHandler mLogFH;

    private static FileLogger instance;

    public static FileLogger getInstance() {
        if (null == instance) {
            instance = new FileLogger();
        }

        return instance;
    }

    public static Logger getLogger() {
        return getInstance().mLoger;
    }

    public FileLogger() {
        mLogTag =  ("P" + System.currentTimeMillis() % 100000);

        String logfn;
        String en = Environment.getExternalStorageState();
        if (en.equals(Environment.MEDIA_MOUNTED)) {
            File sdcardDir = Environment.getExternalStorageDirectory();
            String path = sdcardDir.getPath() + "/CamerajobLogs";
            File path1 = new File(path);
            if (!path1.exists()) {
                path1.mkdirs();
            }

            logfn = path + "/" + LOG_NAME;

        } else {
            File innerPath = ContextUtil.getInstance()
                    .getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS);
            logfn = innerPath.getPath() + "/" + LOG_NAME;
        }

        try {
            mLogFH = new FileHandler(logfn, true);
            mLogFH.setFormatter(new SimpleFormatter() {
                @Override
                public String format(LogRecord record) {

                    String ret = String.format("%s|%s|%s-%d|%s:%s|%s"
                            , UtilFun.MilliSecsToString(record.getMillis())
                            , record.getLevel().getName()
                            , mLogTag ,record.getThreadID()
                            , record.getSourceClassName(), record.getSourceMethodName()
                            , formatMessage(record)) + (System.lineSeparator());
                    return ret;
                }
            });

            mLoger = Logger.getLogger("camerajob_runlog");
            mLoger.addHandler(mLogFH);
            mLoger.setLevel(Level.INFO);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/wxm/camerajob/utility/log/FileLogger.kt
package com.wxm.camerajob.utility.log

import android.util.Log
import com.wxm.camerajob.BuildConfig
import com.wxm.camerajob.utility.AppUtil
import wxm.androidutil.time.toFullTag
import java.io.IOException
import java.sql.Timestamp
import java.util.*
import java.util.logging.*

/**
 * log in file
 * Created by WangXM on 2016/6/18.
 */
class FileLogger private constructor() {
    private val mLogTag: String = "P" + System.currentTimeMillis() % 100000
    private lateinit var mLogger: Logger

    init {
        Log.i(LOG_TAG, "init file logger")

        try {
            FileHandler(
                    "${AppUtil.getLogRootDir()}/${String.format(Locale.CHINA, LOG_NAME, mLogTag)}",
                    true).let {
                it.formatter = object : SimpleFormatter() {
                    override fun format(record: LogRecord): String {
                        return String.format(Locale.CHINA,
                                "%s|%s|%s-%d|%s:%s|%s",
                                Timestamp(record.millis).toFullTag(),
                                record.level.name, mLogTag, record.threadID,
                                record.sourceClassName, record.sourceMethodName,
                                formatMessage(record)) + System.lineSeparator()
                    }
                }

                mLogger = Logger.getLogger("cameraJobRunLog")
                mLogger.addHandler(it)
            }

            @Suppress("ConstantConditionIf")
            mLogger.level = if (BuildConfig.DEBUG) Level.INFO else Level.WARNING
        } catch (e: IOException) {
            Log.e(LOG_TAG, "create file for log failure", e)
        }
    }

    companion object {
        private val LOG_TAG = ::FileLogger.javaClass.simpleName
        private const val LOG_NAME = "cameraJob_run_%s.log"

        private val instance: FileLogger = FileLogger()

        /**
         * get logger to use
         */
        fun getLogger(): Logger {
            return instance.mLogger
        }
    }
}
